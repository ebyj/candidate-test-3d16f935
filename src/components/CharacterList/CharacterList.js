import React from "react";
import PropTypes from "prop-types";
import generateCharacterKey from "./generateKey";

const CharacterPropType = PropTypes.shape({
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  significanceIndex: PropTypes.number.isRequired,
  avatar: PropTypes.string.isRequired,
});

const CharacterListItem = ({ character }) => (
  <div>
    <div>
      <img
        src={`characters/${character.avatar}`}
        alt={character.name}
        loading={"lazy"}
      />
    </div>
    <div>
      <h3 className="character-list-item-name">{character.name}</h3>
      <p className="character-list-item-description">{character.description}</p>
    </div>
  </div>
);

CharacterListItem.propTypes = { character: CharacterPropType };

const CharacterList = ({ characters }) => {
  return (
    <div>
      {characters.map((character, i) => (
        <CharacterListItem
          key={generateCharacterKey(i, character.name)}
          character={character}
        />
      ))}
    </div>
  );
};

CharacterList.propTypes = {
  characters: PropTypes.arrayOf(CharacterPropType).isRequired,
};

export default CharacterList;
