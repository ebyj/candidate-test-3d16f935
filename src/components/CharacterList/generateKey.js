const generateCharacterKey = (index, name) =>
  `${index}-${name.replace(/\s/g, "-")}`;

export default generateCharacterKey;
