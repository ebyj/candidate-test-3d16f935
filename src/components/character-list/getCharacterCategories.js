// Ideally would have a store and selectors to get this kind of information
const getCharacterCategories = (characters) =>
  characters.reduce(
    (acc, character) => {
      if (!acc.includes(character.category)) {
        acc.push(character.category);
      }
      return acc;
    },
    [""]
  );

export default getCharacterCategories;
