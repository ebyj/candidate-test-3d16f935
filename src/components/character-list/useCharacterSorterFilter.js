import { useEffect, useState } from "react";

const useCharacterSorterFilter = (characters, categoryFilter, order) => {
  const [updatedCharacters, setUpdatedCharacters] = useState(characters);

  useEffect(() => {
    const charSlice = characters.slice();
    const updatedCharacters = categoryFilter
      ? charSlice.filter((character) => character.category === categoryFilter)
      : charSlice;

    if (order === "alphabetical") {
      updatedCharacters.sort((a, b) => a.name.localeCompare(b.name));
    } else {
      updatedCharacters.sort(
        (a, b) => a.significanceIndex - b.significanceIndex
      );
    }

    setUpdatedCharacters(updatedCharacters);
  }, [characters, categoryFilter, order]);

  return updatedCharacters;
};

export default useCharacterSorterFilter;
