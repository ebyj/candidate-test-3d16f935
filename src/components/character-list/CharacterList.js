import React, { useMemo, useState } from "react";
import PropTypes from "prop-types";
import generateCharacterKey from "./generateKey";
import styles from "./CharacterList.module.scss";
import { capitalize } from "lodash";
import useCharacterSorterFilter from "./useCharacterSorterFilter";
import getCharacterCategories from "./getCharacterCategories";

const CharacterPropType = PropTypes.shape({
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  significanceIndex: PropTypes.number.isRequired,
  avatar: PropTypes.string.isRequired,
});

const CharacterListItem = ({ character }) => (
  <div className={styles.characterListItem}>
    <img
      className={styles.avatar}
      src={`characters/${character.avatar}`}
      alt={character.name}
    />
    <div className={styles.characterInfo}>
      <div className={styles.characterName}>{character.name}</div>
      <div className={styles.characterCategory}>
        {capitalize(character.category)}
      </div>
      <div className={styles.description}>{character.description}</div>
    </div>
  </div>
);

CharacterListItem.propTypes = { character: CharacterPropType };

const CharacterList = ({ characters }) => {
  const [categoryFilter, setCategoryFilter] = useState("");
  const [order, setOrder] = useState("significanceIndex");

  const sortedFilteredCharacters = useCharacterSorterFilter(
    characters,
    categoryFilter,
    order
  );

  const onCategoryFilterChange = (event) => {
    setCategoryFilter(event.target.value);
  };

  const onOrderAlphabeticallyChange = (event) => {
    setOrder(event.target.value);
  };

  const categories = useMemo(
    () => getCharacterCategories(characters),
    [characters]
  );

  return (
    <div>
      <div className={styles.controls}>
        <div>
          <label htmlFor="character-category-select">Category</label>
          <select
            onChange={onCategoryFilterChange}
            id="character-category-select"
            value={categoryFilter}
          >
            {categories.map((category) => (
              <option key={category} value={category}>
                {capitalize(category || "All")}
              </option>
            ))}
          </select>
        </div>
        <div>
          <label htmlFor="character-order-by">Order by</label>
          <select
            onChange={onOrderAlphabeticallyChange}
            id="character-order-by"
            value={order}
          >
            <option value="significanceIndex">Significance</option>
            <option value="alphabetical">Alphabetical</option>
          </select>
        </div>
      </div>
      <div className={styles.characterList}>
        {sortedFilteredCharacters.map((character, i) => (
          <CharacterListItem
            key={generateCharacterKey(i, character.name)}
            character={character}
          />
        ))}
      </div>
    </div>
  );
};

CharacterList.propTypes = {
  characters: PropTypes.arrayOf(CharacterPropType).isRequired,
};

export default CharacterList;
