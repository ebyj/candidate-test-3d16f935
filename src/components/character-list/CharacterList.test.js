import React from "react";
import useCharacterSorterFilter from "./useCharacterSorterFilter";
import getCharacterCategories from "./getCharacterCategories";
import { act, fireEvent, render, screen } from "@testing-library/react";
import CharacterList from "./CharacterList";

// mock out sorting/filtering hook and character category util
jest.mock("./useCharacterSorterFilter");
jest.mock("./getCharacterCategories");

describe("characterList", () => {
  const characters = [
    {
      name: "Frodo Baggins",
      category: "hobbit",
      description:
        "A young well-to-do hobbit. When he discovers that the magic ring left to him by his eccentric Uncle Bilbo is the One Ring, he reluctantly takes on the quest to destroy it. Often referred to as the Ringbearer.",
      significanceIndex: 0,
      avatar: "frodo_baggins.jpg",
    },
    {
      name: "Gandalf the Grey",
      category: "wizard",
      description:
        "A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.",
      significanceIndex: 1,
      avatar: "gandalf_the_grey.jpg",
    },
    {
      name: "Samwise Gamgee",
      category: "hobbit",
      description:
        "Frodo's gardener at home, and his servant and friend on the quest. Sam is very fond of stories about dragons and elves. A member of the Fellowship, he stays with Frodo after it is broken.",
      significanceIndex: 2,
      avatar: "samwise_gamgee.jpg",
    },
  ];

  beforeEach(() => {
    useCharacterSorterFilter.mockImplementation((characters) => characters);
    getCharacterCategories.mockReturnValue(["", "hobbit", "wizard"]);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders with no characters", () => {
    render(<CharacterList characters={[]} />);

    const characterListContainer = screen.getByTestId("character-list");

    expect(characterListContainer).not.toBeNull();
    expect(characterListContainer.children).toHaveLength(0);
  });

  it("renders characters", () => {
    render(<CharacterList characters={characters} />);

    const characterListContainer = screen.getByTestId("character-list");

    expect(characterListContainer).not.toBeNull();
    expect(characterListContainer.children).toHaveLength(3);

    // seems reasonable to check a snapshot here to ensure expected info
    // is rendered as expected
    expect(characterListContainer.children[0]).toMatchSnapshot();
  });

  it("displays filter options capitalised and includes 'All' option", () => {
    render(<CharacterList characters={characters} />);

    const dropdown = screen.getByTestId("character-category-select");

    expect(dropdown.options[0].textContent).toEqual("All");
    expect(dropdown.options[1].textContent).toEqual("Hobbit");
    expect(dropdown.options[2].textContent).toEqual("Wizard");
  });

  it("updates filter on dropdown category change", () => {
    render(<CharacterList characters={characters} />);

    expect(useCharacterSorterFilter).toHaveBeenCalledTimes(1);
    expect(useCharacterSorterFilter).toHaveBeenNthCalledWith(
      1,
      characters,
      "",
      "significanceIndex"
    );

    const dropdown = screen.getByTestId("character-category-select");

    act(() => {
      // simulate dropdown change
      fireEvent.change(dropdown, { target: { value: "hobbit" } });
    });

    expect(useCharacterSorterFilter).toHaveBeenCalledTimes(2);
    expect(useCharacterSorterFilter).toHaveBeenNthCalledWith(
      2,
      characters,
      "hobbit",
      "significanceIndex"
    );
  });

  it("updates sorter on sort by dropdown change", () => {
    render(<CharacterList characters={characters} />);

    expect(useCharacterSorterFilter).toHaveBeenCalledTimes(1);
    expect(useCharacterSorterFilter).toHaveBeenNthCalledWith(
      1,
      characters,
      "",
      "significanceIndex"
    );

    const dropdown = screen.getByTestId("character-order-by");

    act(() => {
      // simulate dropdown change
      fireEvent.change(dropdown, { target: { value: "alphabetical" } });
    });

    expect(useCharacterSorterFilter).toHaveBeenCalledTimes(2);
    expect(useCharacterSorterFilter).toHaveBeenNthCalledWith(
      2,
      characters,
      "",
      "alphabetical"
    );
  });

  it("only calls getCharacterCategories when characters updates", () => {
    const { rerender } = render(<CharacterList characters={characters} />);

    expect(getCharacterCategories).toHaveBeenCalledTimes(1);

    // manual rerender
    rerender(<CharacterList characters={characters} />);

    expect(getCharacterCategories).toHaveBeenCalledTimes(1);

    // trigger some kind of rerender by updating sort state
    const dropdown = screen.getByTestId("character-order-by");

    act(() => {
      fireEvent.change(dropdown, { target: { value: "alphabetical" } });
    });

    expect(getCharacterCategories).toHaveBeenCalledTimes(1);

    // new array ref should cause getCharacterCategories to be called again
    rerender(<CharacterList characters={[...characters]} />);
    expect(getCharacterCategories).toHaveBeenCalledTimes(2);
  });
});
