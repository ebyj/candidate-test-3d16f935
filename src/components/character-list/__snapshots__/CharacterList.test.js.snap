// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`characterList renders characters 1`] = `
<div
  class="characterListItem"
>
  <img
    alt="Frodo Baggins"
    class="avatar"
    src="characters/frodo_baggins.jpg"
  />
  <div
    class="characterInfo"
  >
    <div
      class="characterName"
    >
      Frodo Baggins
    </div>
    <div
      class="characterCategory"
    >
      Hobbit
    </div>
    <div
      class="description"
    >
      A young well-to-do hobbit. When he discovers that the magic ring left to him by his eccentric Uncle Bilbo is the One Ring, he reluctantly takes on the quest to destroy it. Often referred to as the Ringbearer.
    </div>
  </div>
</div>
`;
